webhook/psbinding/index.go

```diff
@@ -79,3 +83,76 @@ func (im inexactMatcher) Get(key inexactKey, ls labels.Set) (bindable Bindable,
 	}
 	return nil, false
 }
+
+// index is a collection of Bindables indexed by their subject resources.
+type index struct {
+	// lock protects access to exact and inexact
+	lock    sync.RWMutex
+	exact   exactMatcher
+	inexact inexactMatcher
+}
+
+// IndexBuilder allows an index to be built atomically
+type IndexBuilder struct {
+	exact   exactMatcher
+	inexact inexactMatcher
+}
+
+// NewIndexBuilder constructs a new IndexBuilder.
+func NewIndexBuilder() *IndexBuilder {
+	return &IndexBuilder{
+		exact:   make(exactMatcher),
+		inexact: make(inexactMatcher),
+	}
+}
+
+// Associate associates a resource with a given exact key with a given Bindable.
+// TODO: allow multiple Bindables to be associated with the same exact key.
+func (ib *IndexBuilder) Associate(key exactKey, fb Bindable) {
+	ib.exact.Add(key, fb)
+}
+
+// AssociateSelection associates resources with the given inexact key and labels matching the given selector with a given Bindable.
+func (ib *IndexBuilder) AssociateSelection(inexactKey inexactKey, selector labels.Selector, fb Bindable) {
+	ib.inexact.Add(inexactKey, selector, fb)
+}
+
+// Build sets the given index to the built value.
+func (ib *IndexBuilder) Build(index *index) {
+	index.setIndex(ib.exact, ib.inexact)
+}
+
+func (idx *index) setIndex(exact exactMatcher, inexact inexactMatcher) {
+	idx.lock.Lock()
+	defer idx.lock.Unlock()
+	idx.exact = exact
+	idx.inexact = inexact
+}
+
+// LookUp returns the Bindables associated with a resource with the given group, kind, namespace, name, and labels.
+// TODO: find all the matching Bindables instead of just one.
+func (idx *index) LookUp(group string, kind string, namespace string, name string, labels labels.Set) []Bindable {
```

Comment:
```suggestion
func (idx *index) LookUp(group, kind, namespace, name string, labels labels.Set) []Bindable {
```

Reply:
@vagababov Agreed. What I actually did was group the first four parameters into an `exactKey` which makes the call site more readable.